### PR TITLE
[GEOS-10476] STAC Sortables should be the same as the configured queryables

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10746](https://osgeo-org.atlassian.net/browse/GEOS-10746)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/QueryResultBuilder.java [hunk 1]
```diff
@@ -434,14 +434,35 @@ private SortBy[] mapSortProperties(List<String> collectionIds, SortBy[] sortby)
...
             builder = templates.getItemTemplate(null);
+            stacQueryablesBuilder =
+                    new STACQueryablesBuilder(
+                            null,
+                            templates.getItemTemplate(null),
+                            sampleFeatures.getSchema(),
+                            sampleFeatures.getSample(null),
+                            collectionsCache.getCollection(null),
+                            geoServer.getService(OSEOInfo.class));
+            mapper = new STACSortablesMapper(builder, itemsSchema, stacQueryablesBuilder);
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/QueryResultBuilder.java]: fixed

File: src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACService.java [hunk 2]
```diff
@@ -597,14 +597,34 @@ private SortBy[] mapSortProperties(List<String> collectionIds, SortBy[] sortby)
         FeatureType itemsSchema =
                 accessProvider.getOpenSearchAccess().getProductSource().getSchema();
         TemplateBuilder builder;
+        STACSortablesMapper mapper = null;
+        STACQueryablesBuilder stacQueryablesBuilder = null;
         if (collectionIds == null || collectionIds.size() > 1) {
             // right now assuming multiple collections means using search, where the
             // sortables are generic
             builder = templates.getItemTemplate(null);
+            stacQueryablesBuilder =
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACService.java]: fixed

File: src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACService.java [hunk 3]
```diff
@@ -746,7 +766,17 @@ public Sortables collectionSortables(@PathVariable(name = "collectionId") String
         FeatureType itemsSchema =
                 accessProvider.getOpenSearchAccess().getProductSource().getSchema();
         RootBuilder template = this.templates.getItemTemplate(collectionId);
-        Sortables sortables = new STACSortablesMapper(template, itemsSchema, id).getSortables();
+        STACQueryablesBuilder stacQueryablesBuilder =
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/STACService.java]: fixed

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACSortablesMapperTest.java [hunk 4]
```diff
@@ -65,6 +79,50 @@ public void testGetSortable() throws Exception {
...
+
+        // common sortables from spec
+        assertThat(sortables, hasEntry("collection", "parentIdentifier"));
+        assertThat(sortables, hasEntry("datetime", "timeStart"));
+        assertThat(sortables, hasEntry("id", "identifier"));
+
+        assertThat(
+                sortables,
+                not(hasEntry("created", "eop:creationDate"))); // missing because not in the global
+        // queryables
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACSortablesMapperTest.java]: fixed

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACSortablesMapperTest.java [hunk 5]
```diff
@@ -65,6 +79,50 @@ public void testGetSortable() throws Exception {
...
+        assertThat(sortables, not(hasEntry("view:sun_azimuth", "eop:illuminationAzimuthAngle")));
+
+        // one custom, and nested, property
+        assertThat(
+                sortables,
+                hasEntry(
+                        "one.two.three",
+                        "eop:illuminationAzimuthAngle")); // present because it is in the global
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACSortablesMapperTest.java]: fixed

File: src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACSortablesMapperTest.java [hunk 6]
```diff
@@ -65,6 +79,50 @@ public void testGetSortable() throws Exception {
...
+        // common sortables from spec
+        assertThat(sortables, hasEntry("collection", "parentIdentifier"));
+        assertThat(sortables, hasEntry("datetime", "timeStart"));
+        assertThat(sortables, hasEntry("id", "identifier"));
+
+        assertThat(
+                sortables,
+                not(hasEntry("created", "eop:creationDate"))); // missing because not in the global
+        // queryables
+        assertThat(sortables, not(hasEntry("view:sun_azimuth", "eop:illuminationAzimuthAngle")));
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/oseo/oseo-stac/src/test/java/org/geoserver/ogcapi/stac/STACSortablesMapperTest.java:120]: fixed

**Review by aaime — APPROVED** [review]

**Pull request merged**